From 545f7bfa8ea66b64d7251e3d3f886842e87f562f Mon Sep 17 00:00:00 2001
From: Ricky Casavecchia <rickyjames35@gmail.com>
Date: Sat, 13 Jul 2024 23:47:17 +0000
Subject: [PATCH] CMake: Switch to using find_package(Python3) and
 Python3_EXECUTABLE

The old style used was deprecated in CMake 3.12, so this should
not negatively impact our compatibility (requires 3.16)

rpavlik: Only need interpreter, add changelog fragment
---
 CMakeLists.txt                                     | 4 ++--
 changes/sdk/pr.486.gh.OpenXR-SDK-Source.md         | 1 +
 include/openxr/CMakeLists.txt                      | 2 +-
 src/CMakeLists.txt                                 | 8 ++++----
 src/tests/loader_test/test_runtimes/CMakeLists.txt | 2 +-
 5 files changed, 9 insertions(+), 8 deletions(-)
 create mode 100644 changes/sdk/pr.486.gh.OpenXR-SDK-Source.md

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 28eabd1fb..8ed3fe366 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -21,7 +21,7 @@
 cmake_minimum_required(VERSION 3.16)
 project(OPENXR)
 
-find_package(PythonInterp 3)
+find_package(Python3 COMPONENTS Interpreter)
 include(CTest)
 
 # Enable IDE GUI folders.  "Helper targets" that don't have interesting source code should set their FOLDER property to this
@@ -44,7 +44,7 @@ option(
     OFF
 )
 
-if(BUILD_FORCE_GENERATION AND NOT PYTHON_EXECUTABLE)
+if(BUILD_FORCE_GENERATION AND NOT Python3_EXECUTABLE)
     message(FATAL_ERROR "BUILD_FORCE_GENERATION requires Python")
 endif()
 
diff --git a/changes/sdk/pr.486.gh.OpenXR-SDK-Source.md b/changes/sdk/pr.486.gh.OpenXR-SDK-Source.md
new file mode 100644
index 000000000..ecb311cf2
--- /dev/null
+++ b/changes/sdk/pr.486.gh.OpenXR-SDK-Source.md
@@ -0,0 +1 @@
+Improvement: Migrate CMake build system away from using `find_package(PythonInterpreter)`, deprecated since CMake 3.12. Use `find_package(Python3 COMPONENTS Interpreter)` instead.
diff --git a/include/openxr/CMakeLists.txt b/include/openxr/CMakeLists.txt
index b95a60c34..b20af4837 100644
--- a/include/openxr/CMakeLists.txt
+++ b/include/openxr/CMakeLists.txt
@@ -79,7 +79,7 @@ else()
             OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${output}"
             COMMAND
                 "${CMAKE_COMMAND}" -E env "PYTHONPATH=${CODEGEN_PYTHON_PATH}"
-                "${PYTHON_EXECUTABLE}"
+                "${Python3_EXECUTABLE}"
                 "${PROJECT_SOURCE_DIR}/specification/scripts/genxr.py" -registry
                 "${PROJECT_SOURCE_DIR}/specification/registry/xr.xml" -o
                 "${CMAKE_CURRENT_BINARY_DIR}" ${output}
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 94674ced9..470e919ee 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -429,7 +429,7 @@ macro(run_xr_xml_generate dependency output)
         )
         list(APPEND GENERATED_OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/${output}")
     else()
-        if(NOT PYTHON_EXECUTABLE)
+        if(NOT Python3_EXECUTABLE)
             message(
                 FATAL_ERROR
                     "Python 3 not found, but pre-generated ${CMAKE_CURRENT_SOURCE_DIR}/${output} not found"
@@ -439,7 +439,7 @@ macro(run_xr_xml_generate dependency output)
             OUTPUT "${output}"
             COMMAND
                 "${CMAKE_COMMAND}" -E env "PYTHONPATH=${CODEGEN_PYTHON_PATH}"
-                "${PYTHON_EXECUTABLE}"
+                "${Python3_EXECUTABLE}"
                 "${PROJECT_SOURCE_DIR}/src/scripts/src_genxr.py" -registry
                 "${PROJECT_SOURCE_DIR}/specification/registry/xr.xml"
                 "${output}"
@@ -453,7 +453,7 @@ macro(run_xr_xml_generate dependency output)
                 ${ARGN}
             VERBATIM
             COMMENT
-                "Generating ${output} using ${PYTHON_EXECUTABLE} on ${dependency}"
+                "Generating ${output} using ${Python3_EXECUTABLE} on ${dependency}"
         )
         set_source_files_properties(${output} PROPERTIES GENERATED TRUE)
         list(APPEND GENERATED_OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${output}")
@@ -475,7 +475,7 @@ macro(
         OUTPUT "${filename}"
         COMMAND
             "${CMAKE_COMMAND}" -E env "PYTHONPATH=${CODEGEN_PYTHON_PATH}"
-            "${PYTHON_EXECUTABLE}"
+            "${Python3_EXECUTABLE}"
             "${PROJECT_SOURCE_DIR}/src/scripts/generate_api_layer_manifest.py"
             -f "${filename}" -n ${layername} -l ${libfile} -a ${MAJOR}.${MINOR}
             -v ${version} ${genbad} -d ${desc}
diff --git a/src/tests/loader_test/test_runtimes/CMakeLists.txt b/src/tests/loader_test/test_runtimes/CMakeLists.txt
index 67b79c2b6..80117f13e 100644
--- a/src/tests/loader_test/test_runtimes/CMakeLists.txt
+++ b/src/tests/loader_test/test_runtimes/CMakeLists.txt
@@ -49,7 +49,7 @@ macro(gen_xr_runtime_json filename libfile)
     add_custom_command(
         OUTPUT ${filename}
         COMMAND
-            "${PYTHON_EXECUTABLE}"
+            "${Python3_EXECUTABLE}"
             "${PROJECT_SOURCE_DIR}/src/scripts/generate_runtime_manifest.py" -f
             ${filename} -l ${libfile} ${ARGN}
         DEPENDS "${PROJECT_SOURCE_DIR}/src/scripts/generate_runtime_manifest.py"
